### DISCUSSION TITLE
Defining styles for a safe Content Security Policy

### DISCUSSION THREAD

**Discussion by simonduchen** [lit/lit#2576] [General]
Hey Lit Community!

I am working on a project where we are implementing a quite strict [Content Security Policy](https://developers.google.com/web/fundamentals/security/csp) (CSP) which includes not allowing inline-styles. This implies that both `<style></style>` and `<div style=...>` are not allowed.
As the lit documentation describes, the static _styles_ class field allows for performant styling in components which works great with the CSP directives. Though I find it more pleasant and structured to have CSS in an external stylesheet, but that is a subject of taste.

However, for [defining styles in a lit template](https://lit.dev/docs/components/styles/#styles-in-the-template), the recommendation is to not use external stylesheets but instead use inline-styles which makes the website vulnerable to injections and XSS.

### Questions:
As external stylesheets are more safe compared to inline styling.

1. What are your thoughts about using them with the lit framework, regarding both templates and components, and what are the real downsides as of now? 
2. Are there any improvements or new features planned regarding this issue?

**Comment by rictic** [comment]
External style sheets can potentially complicate things, as the URL either needs to be a relative path (in which case, bundlers need to see that URL and rewrite it) or an absolute URL on a public domain (in which case your site isn't hermetic, and the CSS might change independent of your site's code)

That said, disallowing inline-styles is a good option for preventing various content injection attacks site-wide, and if you get a build and dev setup that works well for you it's a good option.


The direction we have been moving to address this and other content-injection style attacks is a combination of using security hooks, allowing a sanitizer library the ability to forbid data bindings based on the HTML position of the binding and the content of the bound data. So e.g. it can forbid binding arbitrary strings into the `style` attribute of an element.

This – combined with enforcing trusted types in CSP and a requirement that trusted type policy creation be gated by the source code that creates the policy being security reviewed – can provide robust security that scales up to large code bases.

**Reply by simonduchen** [reply]
I see, do you have any examples or recommendations of frameworks/libs that work well to provide this kind of scalable security? 

It would be nice to have some kind of documentation or article on lit.dev where you could propose your recommended solutions for using lit together with CSP and in particular this styles issue.

**Reply by justinfagnani** [reply]
We should document this better, since we have a pretty comprehensive set of security features. I opened https://github.com/lit/lit.dev/issues/715 on lit.dev to track.
